File: src/06.hashMap/15.599.ts
export function findRestaurant(list1: string[], list2: string[]): string[] {
    // const length1 = list1.length;
    // const length2 = list2.length;
    // const map = new Map<
    //     string,
    //     {
    //         isPublic: boolean;
    //         index: number;
    //     }
    // >();
    // const res: string[] = [];

    // for (let i = 0; i < length1; i++) {
    //     map.set(list1[i], {
    //         isPublic: false,
    //         index: i,
    //     });
    // }

    // let minIndex = 2000;

    // for (let i = 0; i < length2; i++) {
    //     if (map.has(list2[i])) {
    //         const totalIndex = map.get(list2[i]).index + i;

    //         map.set(list2[i], {
    //             isPublic: true,
    //             index: totalIndex,
    //         });

    //         if (totalIndex <= minIndex) {
    //             minIndex = totalIndex;
    //         }
    //     }
    // }

    // map.forEach((value, key) => {
    //     if (value.isPublic && value.index === minIndex) {
    //         res.push(key);
    //     }
    // });

    // return res;

    // 方法二: 直接在 list2 遍历时, 判断出最小的项
    let res: string[] = [];
    let min = list1.length + list2.length;
    const map = new Map<string, number>();

    for (let i = 0; i < list1.length; i++) {
        map.set(list1[i], i);
    }

    for (let i = 0; i < list2.length; i++) {
        if (!map.has(list2[i])) {
            continue;
        }

        const item = map.get(list2[i]);
        const total = item + i;

        if (total < min) {
            min = total;
            res = [list2[i]];
        } else if (total === min) {
            res.push(list2[i]);
        }
    }

    return res;
}

File: src/06.hashMap/15.599.go
package hashmap

type restaurantMapItem struct {
	isPublic bool
	index    int
}

func FindRestaurantMethod1(list1 []string, list2 []string) []string {
	restaurantMap := map[string]*restaurantMapItem{}
	minIndex := 2000

	for i, v := range list1 {
		restaurantMap[v] = &restaurantMapItem{
			isPublic: false,
			index:    i,
		}
	}

	for i, v := range list2 {
		if value, ok := restaurantMap[v]; ok {
			totalIndex := value.index + i

			restaurantMap[v].isPublic = true
			restaurantMap[v].index = totalIndex

			if totalIndex <= minIndex {
				minIndex = totalIndex
			}
		}
	}

	res := []string{}

	for i, v := range restaurantMap {
		if v.isPublic && v.index == minIndex {
			res = append(res, i)
		}
	}

	return res
}

func FindRestaurant(list1 []string, list2 []string) []string {
	restaurantMap := map[string]int{}
	minIndex := 2000
	res := []string{}

	for i, v := range list1 {
		restaurantMap[v] = i
	}

	for i, v := range list2 {
		if value, ok := restaurantMap[v]; ok {
			totalIndex := value + i

			if totalIndex < minIndex {
				minIndex = totalIndex
				res = []string{v}
			} else if totalIndex == minIndex {
				res = append(res, v)
			}

		}
	}

	return res
}
